File: examples/languages/nodejs/error-handling.ts
/**
 * Error Handling Example
 *
 * Production-ready error handling with:
 * - Automatic retry with exponential backoff
 * - Rate limit handling (429)
 * - Timeout management
 * - Idempotency keys
 * - Custom exceptions
 * - Result validation
 *
 * Usage:
 *   npx ts-node error-handling.ts path/to/image.jpg PATTERN_ID
 */

import axios, { AxiosError, AxiosRequestConfig } from 'axios';
import FormData from 'form-data';
import fs from 'fs';

const IMGGO_API_KEY = process.env.IMGGO_API_KEY;
const IMGGO_BASE_URL = process.env.IMGGO_BASE_URL || 'https://img-go.com/api';

// Custom exceptions
class ImgGoAPIError extends Error {
  constructor(message: string, public statusCode?: number, public response?: any) {
    super(message);
    this.name = 'ImgGoAPIError';
  }
}

class RateLimitError extends ImgGoAPIError {
  constructor(message: string, public retryAfter: number) {
    super(message, 429);
    this.name = 'RateLimitError';
  }
}

class ValidationError extends ImgGoAPIError {
  constructor(message: string) {
    super(message, 400);
    this.name = 'ValidationError';
  }
}

interface JobResponse {
  data: {
    job_id: string;
  };
}

interface JobResult {
  data: {
    status: string;
    manifest?: any;
    result?: any;
    error?: string;
  };
}

async function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function uploadImageWithRetry(
  imagePath: string,
  patternId: string,
  maxRetries: number = 3
): Promise<string> {
  if (!IMGGO_API_KEY) {
    throw new ValidationError('IMGGO_API_KEY environment variable not set');
  }

  if (!fs.existsSync(imagePath)) {
    throw new ValidationError(`Image file not found: ${imagePath}`);
  }

  // Generate idempotency key to prevent duplicate uploads
  const idempotencyKey = `upload-${imagePath}-${patternId}-${Date.now()}`;

  for (let attempt = 0; attempt < maxRetries; attempt++) {
    try {
      console.log(`Upload attempt ${attempt + 1}/${maxRetries}...`);

      const formData = new FormData();
      formData.append('image', fs.createReadStream(imagePath));

      const config: AxiosRequestConfig = {
        headers: {
          'Authorization': `Bearer ${IMGGO_API_KEY}`,
          'Idempotency-Key': idempotencyKey,
          ...formData.getHeaders(),
        },
        timeout: 30000, // 30 second timeout
      };

      const response = await axios.post<JobResponse>(
        `${IMGGO_BASE_URL}/patterns/${patternId}/ingest`,
        formData,
        config
      );

      const jobId = response.data.data.job_id;
      console.log(`Job created: ${jobId}`);
      return jobId;

    } catch (error) {
      const axiosError = error as AxiosError;

      // Handle rate limiting (429)
      if (axiosError.response?.status === 429) {
        const retryAfter = parseInt(axiosError.response.headers['retry-after'] || '60');
        console.log(`Rate limited. Waiting ${retryAfter} seconds...`);

        if (attempt < maxRetries - 1) {
          await sleep(retryAfter * 1000);
          continue;
        }
        throw new RateLimitError(`Rate limit exceeded after ${maxRetries} attempts`, retryAfter);
      }

      // Handle 4xx client errors (don't retry)
      if (axiosError.response?.status && axiosError.response.status >= 400 && axiosError.response.status < 500) {
        const errorMessage = (axiosError.response.data as any)?.error || axiosError.message;
        throw new ValidationError(`Client error (${axiosError.response.status}): ${errorMessage}`);
      }

      // Handle 5xx server errors (retry with exponential backoff)
      if (axiosError.response?.status && axiosError.response.status >= 500) {
        console.error(`Server error (${axiosError.response.status}). Retrying...`);

        if (attempt < maxRetries - 1) {
          const waitTime = Math.pow(2, attempt) * 1000; // Exponential backoff
          console.log(`Waiting ${waitTime / 1000} seconds before retry...`);
          await sleep(waitTime);
          continue;
        }
        throw new ImgGoAPIError(
          `Server error after ${maxRetries} attempts`,
          axiosError.response.status,
          axiosError.response.data
        );
      }

      // Handle connection errors (retry with backoff)
      if (axiosError.code === 'ECONNABORTED' || axiosError.code === 'ETIMEDOUT') {
        console.error('Connection timeout. Retrying...');

        if (attempt < maxRetries - 1) {
          const waitTime = Math.pow(2, attempt) * 1000;
          await sleep(waitTime);
          continue;
        }
        throw new ImgGoAPIError(`Connection timeout after ${maxRetries} attempts`);
      }

      // Unknown error
      if (attempt === maxRetries - 1) {
        throw new ImgGoAPIError(`Upload failed: ${axiosError.message}`);
      }

      // Retry for unknown errors
      const waitTime = Math.pow(2, attempt) * 1000;
      console.log(`Unknown error. Waiting ${waitTime / 1000} seconds before retry...`);
      await sleep(waitTime);
    }
  }

  throw new ImgGoAPIError('Upload failed after all retries');
}

async function pollJobWithRetry(
  jobId: string,
  maxAttempts: number = 60,
  pollInterval: number = 2000
): Promise<any> {
  console.log('Waiting for processing...');

  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    try {
      const response = await axios.get<JobResult>(
        `${IMGGO_BASE_URL}/jobs/${jobId}`,
        {
          headers: { 'Authorization': `Bearer ${IMGGO_API_KEY}` },
          timeout: 10000, // 10 second timeout
        }
      );

      const { status, manifest, result, error } = response.data.data;
      console.log(`  Attempt ${attempt + 1}/${maxAttempts}: ${status}`);

      if (status === 'succeeded') {
        const extractedData = manifest || result;

        // Validate result
        if (!extractedData) {
          throw new ValidationError('Job succeeded but no data returned');
        }

        return extractedData;
      }

      if (status === 'failed') {
        throw new ImgGoAPIError(`Job failed: ${error}`);
      }

      await sleep(pollInterval);

    } catch (error) {
      const axiosError = error as AxiosError;

      // Re-throw custom errors
      if (error instanceof ImgGoAPIError) {
        throw error;
      }

      // Handle timeout during polling (retry without counting against maxAttempts)
      if (axiosError.code === 'ECONNABORTED' || axiosError.code === 'ETIMEDOUT') {
        console.log('  Polling timeout, retrying...');
        await sleep(pollInterval);
        continue;
      }

      // For other errors during polling, log and retry
      console.error(`  Polling error: ${axiosError.message}`);
      await sleep(pollInterval);
    }
  }

  throw new ImgGoAPIError(`Job timeout after ${maxAttempts * pollInterval / 1000} seconds`);
}

async function processImageWithRetry(imagePath: string, patternId: string): Promise<any> {
  const jobId = await uploadImageWithRetry(imagePath, patternId);
  return await pollJobWithRetry(jobId);
}

async function main() {
  const [imagePath, patternId] = process.argv.slice(2);

  if (!imagePath || !patternId) {
    console.log('Usage: npx ts-node error-handling.ts <image_path> <pattern_id>');
    process.exit(1);
  }

  try {
    const result = await processImageWithRetry(imagePath, patternId);
    console.log('\n' + '='.repeat(60));
    console.log('EXTRACTED DATA');
    console.log('='.repeat(60));
    console.log(JSON.stringify(result, null, 2));
  } catch (error: any) {
    console.error(`\nError: ${error.message}`);

    if (error instanceof RateLimitError) {
      console.error(`Please wait ${error.retryAfter} seconds before retrying.`);
    }

    if (error.statusCode) {
      console.error(`Status Code: ${error.statusCode}`);
    }

    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

export { processImageWithRetry, uploadImageWithRetry, pollJobWithRetry };

File: examples/languages/python/error-handling.py
"""
Error Handling and Retry Logic

Production-ready example with proper error handling,
exponential backoff, and retry logic.

Usage:
    python error-handling.py path/to/image.jpg PATTERN_ID
"""

import os
import sys
import time
import requests
from typing import Optional, Dict, Any


# Configuration
IMGGO_API_KEY = os.getenv("IMGGO_API_KEY")
IMGGO_BASE_URL = os.getenv("IMGGO_BASE_URL", "https://img-go.com/api")


class ImgGoError(Exception):
    """Base exception for ImgGo API errors"""
    pass


class RateLimitError(ImgGoError):
    """Raised when rate limit is exceeded"""
    pass


class JobFailedError(ImgGoError):
    """Raised when job processing fails"""
    pass


def upload_with_retry(
    image_path: str,
    pattern_id: str,
    max_retries: int = 3
) -> dict:
    """
    Upload an image with automatic retry on failure.

    Args:
        image_path: Path to the image file
        pattern_id: Pattern ID to use
        max_retries: Maximum number of retry attempts

    Returns:
        dict: Extracted data

    Raises:
        ImgGoError: If all retries fail
    """
    if not IMGGO_API_KEY:
        raise ValueError("IMGGO_API_KEY environment variable not set")

    for attempt in range(max_retries):
        try:
            print(f"Upload attempt {attempt + 1}/{max_retries}")

            # Upload image
            import mimetypes
            mime_type, _ = mimetypes.guess_type(image_path)

            with open(image_path, 'rb') as f:
                response = requests.post(
                    f"{IMGGO_BASE_URL}/patterns/{pattern_id}/ingest",
                    headers={
                        "Authorization": f"Bearer {IMGGO_API_KEY}",
                        "Idempotency-Key": f"upload-{image_path}-{pattern_id}"
                    },
                    files={"image": (os.path.basename(image_path), f, mime_type or 'image/jpeg')},
                    timeout=30
                )

            # Handle different HTTP status codes
            if response.status_code == 429:
                # Rate limit - wait and retry
                retry_after = int(response.headers.get("Retry-After", 60))
                print(f"Rate limited. Waiting {retry_after} seconds...")
                time.sleep(retry_after)
                continue

            elif response.status_code >= 500:
                # Server error - retry with exponential backoff
                wait_time = 2 ** attempt
                print(f"Server error. Waiting {wait_time} seconds before retry...")
                time.sleep(wait_time)
                continue

            elif response.status_code >= 400:
                # Client error - don't retry
                error_msg = response.json().get("error", {}).get("message", "Unknown error")
                raise ImgGoError(f"Client error: {error_msg}")

            # Success
            response.raise_for_status()
            data = response.json()
            job_id = data["data"]["job_id"]

            print(f"Job created: {job_id}")

            # Poll for results
            result = poll_with_retry(job_id)
            return result

        except requests.exceptions.Timeout:
            print(f"Request timeout. Retrying...")
            if attempt < max_retries - 1:
                time.sleep(2 ** attempt)
            else:
                raise ImgGoError("Max retries exceeded due to timeouts")

        except requests.exceptions.ConnectionError:
            print(f"Connection error. Retrying...")
            if attempt < max_retries - 1:
                time.sleep(2 ** attempt)
            else:
                raise ImgGoError("Max retries exceeded due to connection errors")

    raise ImgGoError("Max retries exceeded")


def poll_with_retry(job_id: str, max_attempts: int = 60) -> dict:
    """
    Poll a job with error handling.

    Args:
        job_id: Job ID to poll
        max_attempts: Maximum polling attempts

    Returns:
        dict: Extracted data

    Raises:
        JobFailedError: If job processing fails
    """
    for attempt in range(max_attempts):
        try:
            response = requests.get(
                f"{IMGGO_BASE_URL}/jobs/{job_id}",
                headers={
                    "Authorization": f"Bearer {IMGGO_API_KEY}"
                },
                timeout=10
            )

            if response.status_code == 404:
                # Job not found - might be temporary, retry
                print(f"  Job not found yet, retrying...")
                time.sleep(2)
                continue

            response.raise_for_status()
            data = response.json()["data"]

            status = data["status"]
            print(f"  Status: {status} ({attempt + 1}/{max_attempts})")

            if status == "succeeded":
                result = data.get("manifest") or data.get("result")

                # Validate result
                if not result:
                    raise JobFailedError("Job succeeded but no result returned")

                return result

            elif status == "failed":
                error = data.get("error", "Unknown error")
                raise JobFailedError(f"Job failed: {error}")

            # Still processing, wait
            time.sleep(2)

        except requests.exceptions.Timeout:
            print(f"  Poll timeout, retrying...")
            time.sleep(1)

        except requests.exceptions.ConnectionError:
            print(f"  Connection error, retrying...")
            time.sleep(2)

    raise ImgGoError(f"Job timeout after {max_attempts} attempts")


def validate_result(result: Any) -> bool:
    """
    Validate extracted data.

    Args:
        result: The extracted data to validate

    Returns:
        bool: True if valid, False otherwise
    """
    if not result:
        return False

    # Add custom validation logic here
    # For example, check required fields exist

    return True


def main():
    """Main entry point with comprehensive error handling"""
    if len(sys.argv) < 3:
        print("Usage: python error-handling.py <image_path> <pattern_id>")
        sys.exit(1)

    image_path = sys.argv[1]
    pattern_id = sys.argv[2]

    if not os.path.exists(image_path):
        print(f"Error: Image file not found: {image_path}")
        sys.exit(1)

    try:
        result = upload_with_retry(image_path, pattern_id)

        # Validate result
        if not validate_result(result):
            print("Warning: Result validation failed")

        print("\n" + "="*60)
        print("EXTRACTED DATA")
        print("="*60)
        print(result)
        print("\nSuccess!")

    except ImgGoError as e:
        print(f"\nAPI Error: {e}")
        sys.exit(1)

    except Exception as e:
        print(f"\nUnexpected Error: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()
